Tornado 6.0 and Jupyter notebook server issue.

We're seeing tests fail in inside of Jupyter(Lab) when using Tornado **6.x**, forcing us to pin to Tornado 5.x. (See the relevant [issue](https://github.com/jupyterlab/jupyterlab/issues/6131)).

Here's the traceback we see in the tests:  
```
    Traceback (most recent call last):
      File "~/dev/tornado/tornado/web.py", line 1699, in _execute
        result = await result
      File "~/dev/tornado/tornado/gen.py", line 736, in run
        yielded = self.gen.throw(*exc_info)  # type: ignore
      File "~/dev/notebook/notebook/services/kernels/handlers.py", line 239, in get
        yield super(ZMQChannelsHandler, self).get(kernel_id=kernel_id)
      File "~/dev/tornado/tornado/gen.py", line 729, in run
        value = future.result()
      File "~/dev/tornado/tornado/gen.py", line 736, in run
        yielded = self.gen.throw(*exc_info)  # type: ignore
      File "~/dev/notebook/notebook/base/zmqhandlers.py", line 293, in get
        yield maybe_future(res)
      File "~/dev/tornado/tornado/gen.py", line 729, in run
        value = future.result()
      File "~/dev/tornado/tornado/websocket.py", line 278, in get
        await self.ws_connection.accept_connection(self)
      File "~/dev/tornado/tornado/websocket.py", line 881, in accept_connection
        await self._accept_connection(handler)
      File "~/dev/tornado/tornado/websocket.py", line 964, in _accept_connection
        await self._receive_frame_loop()
      File "~/dev/tornado/tornado/websocket.py", line 1118, in _receive_frame_loop
        await self._receive_frame()
      File "~/dev/tornado/tornado/websocket.py", line 1130, in _receive_frame
        data = await self._read_bytes(2)
      File "~/dev/tornado/tornado/websocket.py", line 1124, in _read_bytes
        data = await self.stream.read_bytes(n)
      File "~/dev/tornado/tornado/iostream.py", line 436, in read_bytes
        future = self._start_read()
      File "~/dev/tornado/tornado/iostream.py", line 797, in _start_read
        assert self._read_future is None, "Already reading"
```
I'm hoping y'all might be able to provide some insight on the issue. I can't tell if we missed something in the jump to 6.x that you might be able to answer easily. 😄 

We've tracked down the [exact commit](https://github.com/tornadoweb/tornado/commit/a55c7fe7abcfd3e11cf47efb7759d680a2dac35d) that broke our tests. 

I dove a bit deeper into Tornado and found a single change that can toggle the break. Here's a summary of the change:
* Before this commit, `_maybe_add_error_listener` ensured that all futures+callbacks were done+closed if the stream was closed by calling the`_maybe_run_close_callback()` method 
* `_maybe_run_close_callback()` was renamed to `_signal_closed()`
* In this commit, the `_maybe_add_error_listener` no longer calls `signal_closed`/`_maybe_run_close_callback()`

Reverting to previous behavior fixes the issue. Here's a patch showing the [diff](https://github.com/Zsailer/tornado/commit/60d67236e736973a96b9d628baa3bc6f628fa1b1). I add this `_signal_closed()` method back to `_maybe_add_error_listener`    and things seem to work again.

I'm not sure if this is a bug in Tornado, or if Jupyter's notebook server is using this API wrong (possibly a missing `await` could be a likely culprit). 

Any thoughts? Let me know if I can provide you more detail.

If I may throw in another backtrace here which may be related:

```
  File "/usr/local/lib/python3.6/dist-packages/tornado/http1connection.py", line 735, in _server_request_loop
    ret = yield conn.read_response(request_delegate)
  File "/usr/local/lib/python3.6/dist-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/local/lib/python3.6/dist-packages/tornado/gen.py", line 326, in wrapper
    yielded = next(result)
  File "/usr/local/lib/python3.6/dist-packages/tornado/http1connection.py", line 160, in _read_message
    max_bytes=self.params.max_header_size)
  File "/usr/local/lib/python3.6/dist-packages/tornado/iostream.py", line 353, in read_until_regex
    future = self._set_read_callback(callback)
  File "/usr/local/lib/python3.6/dist-packages/tornado/iostream.py", line 860, in _set_read_callback
    assert self._read_future is None, "Already reading"
AssertionError: Already reading
```

This is from tornado-5.1.1 on python-3.6. We have a couple thousand long-lived websocket connections to each websocket-server process (and they also serve a few plain http requests per minute). This seems to be correlated to some cases where a websocket seems to already be closed when our (plain sychronous) `open()` method calls `self.write_message("ACK")`. With about 5k clients spread across 3 server processes, we see the close-during-open about 40 times a day, and this backtrace just a couple times a week.

I'm pretty sure this never happened with tornado-4.5.3. It's no big problem currently ... it's probably the fault of some distant client in some very strange network conditions. It's hard for me to tell whether something like #2652 would fix this occasional exception for me, but if it seems like it should, I would be interested in trying to backport it to 5.1.

--- a/tornado/iostream.py
+++ b/tornado/iostream.py
@@ -794,6 +794,7 @@ class BaseIOStream(object):
             self._read_from_buffer(pos)
 
     def _start_read(self) -> Future:
+        self._check_closed() # Before reading, check that stream is not closed.
         assert self._read_future is None, "Already reading"
         self._read_future = Future()
         return self._read_future

--- a/tornado/iostream.py
+++ b/tornado/iostream.py
@@ -794,7 +794,7 @@ class BaseIOStream(object):
             self._read_from_buffer(pos)
 
     def _start_read(self) -> Future:
-        self._check_closed() # Before reading, check that stream is not closed.
+        self._check_closed()  # Before reading, check that stream is not closed.
         assert self._read_future is None, "Already reading"
         self._read_future = Future()
         return self._read_future

